# Predicting Average House Prices in Madrid Neighbourhoods Using Socioeconomic Indicators

## Overall Project Aim:
To develop and evaluate a machine learning model that uses 2025 socioeconomic indicators to predict 2025 average housing prices across Madrid neighbourhoods.

## 1. Data Inspection

### Introduction
Two datasets are used in this project:

- Panel of Socioeconomic Indicators of Districts and Neighbourhoods of Madrid (indicators.csv)
- Madrid Average Declared Housing Price (€ per m²) by District and Neighbourhood (prices.csv)

This notebook focuses on exploring the raw Madrid average housing price dataset. The dataset contains average housing prices for new, used and total housing across multiple years, with data available at both district and neighbourhood level.

### Import Libraries

In [1]:
# Import the Pandas library for data manipulation and analysis
import pandas as pd

### Load Dataset

In [2]:
# Loading Dataset 2: Average Declared Housing Price (euros/m2) by District and Neighborhood in Madrid
prices = pd.read_csv("../data/prices.csv", sep=";",skiprows=5)

### Observations

After the dataset initially failed to load correctly, inspection of prices.csv showed that the first five rows contained metadata. These rows were therefore skipped when loading the dataset.

### Initial Inspection

In [3]:
# Display the first ten rows
prices.head(10)

,Unnamed: 0,Unnamed: 1,2007,2007.1,2007.2,2007.3,2007.4,2007.5,2007.6,2007.7,...,2024.8,2025,2025.1,2025.2,2025.3,2025.4,2025.5,2025.6,2025.7,2025.8
0,NaN,NaN,Euros/m2,Euros/m2,Euros/m2,Números índice (Madrid=100),Números índice (Madrid=100),Números índice (Madrid=100),Números índice (Total=100),Números índice (Total=100),...,Números índice (Total=100),Euros/m2,Euros/m2,Euros/m2,Números índice (Madrid=100),Números índice (Madrid=100),Números índice (Madrid=100),Números índice (Total=100),Números índice (Total=100),Números índice (Total=100)
1,NaN,NaN,Total,Nuevas,Usadas,Total,Nuevas,Usadas,Total,Nuevas,...,Usadas,Total,Nuevas,Usadas,Total,Nuevas,Usadas,Total,Nuevas,Usadas
2,Ciudad de Madrid,Ciudad de Madrid,"3.746,27","3.479,98","3.859,89",100,100,100,100,"92,89",...,"100,39","5.285,72","5.041,46","5.333,59",100,100,100,100,"95,38","100,91"
3,01. Centro,01. Centro,"4.798,92","4.988,4","4.768,62","128,1","143,35","123,54",100,"103,95",...,"100,18","7.147,7","6.814,13","7.171,36","135,23","135,16","134,46",100,"95,33","100,33"
4,01. Centro,011. Palacio,..,..,..,..,..,..,..,..,...,"99,32","6.895,15","6.421,28","6.954,58","130,45","127,37","130,39",100,"93,13","100,86"
5,01. Centro,012. Embajadores,..,..,..,..,..,..,..,..,...,"99,95","6.054,75","6.153,41","6.044,57","114,55","122,06","113,33",100,"101,63","99,83"
6,01. Centro,013. Cortes,..,..,..,..,..,..,..,..,...,"99,71","7.246,75",0,"7.246,75","137,1",0,"135,87",100,0,100
7,01. Centro,014. Justicia,..,..,..,..,..,..,..,..,...,100,"9.541,9","9.505,61","9.544,48","180,52","188,55","178,95",100,"99,62","100,03"
8,01. Centro,015. Universidad,..,..,..,..,..,..,..,..,...,"99,6","7.135,43","6.682,45","7.149,96","134,99","132,55","134,06",100,"93,65","100,2"
9,01. Centro,016. Sol,..,..,..,..,..,..,..,..,...,"101,95","7.696,61",0,"7.726,49","145,61",0,"144,86",100,0,"100,39"


In [4]:
# Display the last five rows
prices.tail()

,Unnamed: 0,Unnamed: 1,2007,2007.1,2007.2,2007.3,2007.4,2007.5,2007.6,2007.7,...,2024.8,2025,2025.1,2025.2,2025.3,2025.4,2025.5,2025.6,2025.7,2025.8
154,21. Barajas,215. Corralejos,..,..,..,..,..,..,..,..,...,"102,26","4.890,3",0,"4.902,36","92,52",0,"91,91",100,0,"100,25"
155,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
156,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
157,Fuente: Colegio de Registradores de España,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
158,Observaciones: Tasa de cobertura: La informaci...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [5]:
# Display ten random rows
prices.sample(10)

,Unnamed: 0,Unnamed: 1,2007,2007.1,2007.2,2007.3,2007.4,2007.5,2007.6,2007.7,...,2024.8,2025,2025.1,2025.2,2025.3,2025.4,2025.5,2025.6,2025.7,2025.8
124,16. Hortaleza,166. Valdefuentes,..,..,..,..,..,..,..,..,...,96,"5.977,26","5.655,9","6.072,64","113,08","112,19","113,86",100,"94,62","101,6"
100,13. Puente de Vallecas,136. Numancia,..,..,..,..,..,..,..,..,...,"98,64","3.239,46","3.122,09","3.257,48","61,29","61,93","61,07",100,"96,38","100,56"
113,15. Ciudad Lineal,155. San Pascual,..,..,..,..,..,..,..,..,...,"99,13","5.566,9",0,"5.566,9","105,32",0,"104,37",100,0,100
138,19. Vicálvaro,193. Valderrivas,-,-,-,-,-,-,-,-,...,"101,37","3.546,72",0,"3.547,8","67,1",0,"66,52",100,0,"100,03"
145,20. San Blas-Canillejas,205. Rosas,..,..,..,..,..,..,..,..,...,"100,79","4.493,65",0,"4.534,04","85,01",0,"85,01",100,0,"100,9"
15,02. Arganzuela,025. Delicias,..,..,..,..,..,..,..,..,...,"100,14","5.426,01",0,"5.499,92","102,65",0,"103,12",100,0,"101,36"
154,21. Barajas,215. Corralejos,..,..,..,..,..,..,..,..,...,"102,26","4.890,3",0,"4.902,36","92,52",0,"91,91",100,0,"100,25"
91,12. Usera,125. Moscardó,..,..,..,..,..,..,..,..,...,"98,8","3.739,18","4.123,27","3.701,66","70,74","81,79","69,4",100,"110,27",99
111,15. Ciudad Lineal,153. Quintana,..,..,..,..,..,..,..,..,...,"99,33","4.436,01",0,"4.420,93","83,92",0,"82,89",100,0,"99,66"
93,12. Usera,127. Pradolongo,..,..,..,..,..,..,..,..,...,0,"3.390,16",0,"3.333,54","64,14",0,"62,5",100,0,0


In [6]:
# Check the number of rows and columns
prices.shape

(159, 173)

In [7]:
# Display all column names
prices.columns

Index(['Unnamed: 0', 'Unnamed: 1', '2007', '2007.1', '2007.2', '2007.3',
       '2007.4', '2007.5', '2007.6', '2007.7',
       ...
       '2024.8', '2025', '2025.1', '2025.2', '2025.3', '2025.4', '2025.5',
       '2025.6', '2025.7', '2025.8'],
      dtype='str', length=173)

### Observations

The dataset contains multi-year data at both district and neighbourhood level. Each row represents either the city of Madrid, a district or a neighbourhood, while the columns contain the corresponding average housing price in euros per square metre for new, used and total housing between 2007 and 2025.

While some metadata was removed when loading the dataset, the column titles remain split across the header and first two rows. Additionally, the final four rows (155–158) appear to contain metadata rather than housing price observations. The data therefore requires further removal of metadata, restructuring and filtering by year, housing type and geographical level to obtain the target variable for eventual modelling.

### Data Types

In [8]:
# Display the data type of each column
prices.dtypes

Unnamed: 0    str
Unnamed: 1    str
2007          str
2007.1        str
2007.2        str
             ... 
2025.4        str
2025.5        str
2025.6        str
2025.7        str
2025.8        str
Length: 173, dtype: object

In [9]:
# Display dataset information
prices.info()

<class 'pandas.DataFrame'>
RangeIndex: 159 entries, 0 to 158
Columns: 173 entries, Unnamed: 0 to 2025.8
dtypes: str(173)
memory usage: 327.0 KB


### Observations
The dataset contains 173 columns, which are initially stored as a string data type. As this project focuses specifically on 2025 data, the relevant housing price values will need to be converted to a numerical data type (e.g float) before modelling.

Additionally, the values use European number formatting, with a comma (,) as the decimal separator and a full stop (.) as the thousands separator. These values will therefore need to be reformatted to ensure they are interpreted correctly as numerical data.

In [10]:
# Summary statistics
prices.describe()

,Unnamed: 0,Unnamed: 1,2007,2007.1,2007.2,2007.3,2007.4,2007.5,2007.6,2007.7,...,2024.8,2025,2025.1,2025.2,2025.3,2025.4,2025.5,2025.6,2025.7,2025.8
count,155,153,155,155,155,155,155,155,155,155,...,155,155,155,155,155,155,155,155,155,155
unique,24,153,26,26,26,26,26,26,5,26,...,133,155,94,154,155,94,154,3,94,125
top,15. Ciudad Lineal,Ciudad de Madrid,..,..,..,..,..,..,..,..,...,100,Euros/m2,0,0,Números índice (Madrid=100),0,0,100,0,100
freq,10,1,128,128,128,128,128,128,128,128,...,10,1,62,2,1,62,2,153,62,8


In [11]:
# Summary statistics for all columns
prices.describe(include="all")

,Unnamed: 0,Unnamed: 1,2007,2007.1,2007.2,2007.3,2007.4,2007.5,2007.6,2007.7,...,2024.8,2025,2025.1,2025.2,2025.3,2025.4,2025.5,2025.6,2025.7,2025.8
count,155,153,155,155,155,155,155,155,155,155,...,155,155,155,155,155,155,155,155,155,155
unique,24,153,26,26,26,26,26,26,5,26,...,133,155,94,154,155,94,154,3,94,125
top,15. Ciudad Lineal,Ciudad de Madrid,..,..,..,..,..,..,..,..,...,100,Euros/m2,0,0,Números índice (Madrid=100),0,0,100,0,100
freq,10,1,128,128,128,128,128,128,128,128,...,10,1,62,2,1,62,2,153,62,8


### Missing Values and Duplicates

In [12]:
# Count missing values
prices.isnull().sum()

Unnamed: 0    4
Unnamed: 1    6
2007          4
2007.1        4
2007.2        4
             ..
2025.4        4
2025.5        4
2025.6        4
2025.7        4
2025.8        4
Length: 173, dtype: int64

In [13]:
# Count duplicate rows
prices.duplicated().sum()

np.int64(1)

In [14]:
# Count number of unique values in each column
prices.nunique()

Unnamed: 0     24
Unnamed: 1    153
2007           26
2007.1         26
2007.2         26
             ... 
2025.4         94
2025.5        154
2025.6          3
2025.7         94
2025.8        125
Length: 173, dtype: int64

### Observations 

Initial summary statistics are of limited value at this stage, as the data has not yet been cleaned. Missing values indicate that not every observation contains a value for each corresponding column. Inspection using tail() also shows that the final four rows contain information outside of the main dataset, which may account for some of the identified missing values.

One duplicate row was also identified during the initial inspection. This may be associated with the metadata contained within the final rows and will be investigated during data cleaning.